From a8dc28f123b4b827e737ed35a50290eba7084a1d Mon Sep 17 00:00:00 2001
From: David Howells <dhowells@redhat.com>
Date: Thu, 27 Nov 2014 11:46:57 -0500
Subject: [fs] ovl: use lockless_dereference() for upperdentry

Message-id: <20141127114657.9170.25331.stgit@warthog.procyon.org.uk>
Patchwork-id: 100988
O-Subject: [RHEL7 PATCH 5/7] ovl: use lockless_dereference() for upperdentry
Bugzilla: 985875
RH-Acked-by: Alexander Viro <aviro@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>

From: Miklos Szeredi <mszeredi@suse.cz>

Don't open code lockless_dereference() in ovl_upperdentry_dereference().

Signed-off-by: Miklos Szeredi <mszeredi@suse.cz>
Original-commit: 71d509280f7e92eb60ae6b7c78c20afafff060c7
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/overlayfs/super.c b/fs/overlayfs/super.c
index 74fa03b..24a7fe3 100644
--- a/fs/overlayfs/super.c
+++ b/fs/overlayfs/super.c
@@ -84,12 +84,7 @@ enum ovl_path_type ovl_path_type(struct dentry *dentry)
 
 static struct dentry *ovl_upperdentry_dereference(struct ovl_entry *oe)
 {
- struct dentry *upperdentry = ACCESS_ONCE(oe->__upperdentry);
- /*
-  * Make sure to order reads to upperdentry wrt ovl_dentry_update()
-  */
- smp_read_barrier_depends();
- return upperdentry;
+ return lockless_dereference(oe->__upperdentry);
 }
 
 void ovl_path_upper(struct dentry *dentry, struct path *path)
-- 
1.7.1